doc/sfp-refactor/sfp-refactor.md

```diff
@@ -0,0 +1,816 @@
+
+#### Regular Identifier
+
+1. Lookup “public” XcvrMemMap
+2. Initialize XcvrMemMap
+3. Retrieve vendor ID via XcvrMemMap
+4. Lookup mapping for vendor-specific definitions based on vendor ID
+5. If vendor-specific classes are found, use them instead  
+6. Else, proceed with currently initialized classes
+
+
+#### Vendor Specific Identifier
+
+1. Look up possible XcvrMemMaps associated with the identifier
+2. For each XcvrMemMap found
+    1. Initialize XcvrMemMap
+    2. Retrieve vendor ID via XcvrMemMap
+    3. Compare retrieved vendor ID with expected vendor ID (according to the mapping definition)
+    4. If match, we’re done; proceed with currently initialized XcvrMemMap
+    5. Else, continue with next XcvrMemMap
+
+
+## SfpBase Modifications
+
+SfpBase will gain two new methods in the new design: refresh_xcvr_api() and get_xcvr_api(). Both will be called by xcvrd.
+
+```
+# sonic-platform-common/sonic_platform_base/sfp_base.py
+
+class SfpBase(device_base.DeviceBase):
+    def __init__(self):
+        self.xcvr_api_factory = XcvrApiFactory(self.read_eeprom, self.write_eeprom)
+        self.xcvr_api = None
+    ...
+
+    def refresh_xcvr_api(self):
+        """
+        Updates the XcvrApi associated with this SFP
+        """
+
+    def get_xcvr_api(self):
+        """
+        Retrieves the XcvrApi associated with this SFP
+
+        Returns:
+            An object derived from XcvrApi that corresponds to the SFP
+        """
+    ... 
+
+```
+### SfpOptoeBase
+
+The long-term plan for this refactor is to have SfpBase primarily be an interface for functionality that is expected to vary across platforms and thus would need to be implemented by platform vendors. This means that methods like get_temperature, get_transceiver_info(), etc. from SfpBase that are xcvr EEPROM-related and whose implementations are not actually platform dependent will get moved to XcvrApi.
+
+In the short-term, the methods will be kept in SfpBase. A new class, SfpOptoeBase, will inherit SfpBase and implement these methods simply by deferring to XcvrApi. Platform vendors that wish to make use of SfpOptoeBase will create an Sfp class that inherits from SfpOptoeBase rather than SfpBase; this is illustrated below by platform vendor A and B’s Sfp classes. Vendors that continue inheriting from SfpBase (vendors C and D) can continue to do so without being affected negatively. 
+
+![](../../images/sfp-refactor/sfp-optoe-base-diagram.png)
+
+SfpOptoeBase will also implement read_eeprom and write_eeprom, since for platforms that rely on optoe for xcvr EEPROM access the only aspect that is platform dependent is the mapping between front panel ports and the EEPROM file path; for this a new method, get_eeprom_path, will need to be implemented.
+
+
+```
+# sfp_optoe_base.py
+
+class SfpOptoeBase(SfpBase):
+    def __init__(self):
+        SfpBase.__init__(self)
+    ...
+
+    def get_rx_los(self):
```

Comment:
They are all in XcvrApi. I'm putting these methods in SfpOptoeBase for now because, as stated in this HLD, these PI methods will stay in SfpBase in the short term and implementing them here in SfpOptoeBase (by deferring to XcvrApi) provides convenience for platforms wishing to migrate to the new design.